internal/docker/imagesgc.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func NewImagesGC(path string) (*ImagesGC, error) {
+	d, err := os.Open(path)
+	if errors.Is(err, fs.ErrNotExist) {
+		gc := defaultImagesGC()
+		gc.path = path
+		return &gc, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer d.Close()
+	dec := json.NewDecoder(d)
+	var entries []gcEntry
+	if err := dec.Decode(&entries); err != nil {
+		return nil, err
+	}
+
+	gc := defaultImagesGC()
+	gc.path = path
+	gc.images = entries
+	return &gc, nil
+}
+
+// Persist saves the list of images to disk.
+func (gc *ImagesGC) Persist() error {
+	if gc.path == "" {
+		return errors.New("GC list was not created with a path")
+	}
+	if len(gc.images) == 0 {
+		return nil
+	}
+
+	d, err := json.Marshal(gc.images)
+	if err != nil {
+		return fmt.Errorf("failed to encode list of images: %w", err)
+	}
+	return os.WriteFile(gc.path, d, 0644)
+}
+
+// Track images before they are downloaded. Images already present are ignored if they are not already tracked.
+func (gc *ImagesGC) Track(images ...string) error {
+	present, err := gc.client.ListImages()
+	if err != nil {
+		return fmt.Errorf("failed to list local Docker images: %w", err)
+	}
+
+	now := gc.clock()
+	for _, image := range images {
+		currentIndex := slices.IndexFunc(gc.images, func(i gcEntry) bool { return i.ImageTag == image })
+		if slices.Contains(present, image) && currentIndex < 0 {
+			// We don't track images already present.
+			continue
+		}
+
+		if currentIndex < 0 {
+			gc.images = append(gc.images, gcEntry{
+				ImageTag: image,
+				LastUsed: now,
+			})
+			continue
+		}
+
+		gc.images[currentIndex].LastUsed = now
+	}
+
+	return nil
+}
+
+// Run runs garbage collection, it removes images according to the rules.
+func (gc *ImagesGC) Run() error {
+	if !gc.Enabled {
+		return nil
+	}
+
+	present, err := gc.client.ListImages()
+	if err != nil {
+		return fmt.Errorf("failed to list local Docker images: %w", err)
+	}
+
+	sizeOk := gc.MaxTotalSize == 0
+	maxUnused := gc.clock().Add(-gc.MaxUnused)
+	var images []gcEntry
+	slices.SortFunc(gc.images, func(a, b gcEntry) int { return a.LastUsed.Compare(b.LastUsed) })
+	for i, image := range gc.images {
+		if !sizeOk {
+			totalSize, err := gc.client.TotalImagesSize()
+			if err != nil {
+				gc.images = append(images, gc.images[i:]...)
+				return fmt.Errorf("cannot get total images size: %w", err)
+			}
+			sizeOk = totalSize <= gc.MaxTotalSize
+		}
+		if !sizeOk || image.LastUsed.Before(maxUnused) {
+			if slices.Contains(present, image.ImageTag) {
+				err := gc.client.RemoveImage(image.ImageTag)
+				if errors.Is(err, BusyImageErr) {
+					continue
+				}
+				if err != nil {
+					gc.images = append(images, gc.images[i:]...)
+					return fmt.Errorf("cannot remove image %s: %w", image.ImageTag, err)
+				}
+				continue
+			}
+		}
+
+		images = append(images, image)
+	}
+
+	gc.images = images
+	return nil
+}
+
+type localImagesGCClient struct {
+}
+
+func defaultImagesGCClient() *localImagesGCClient {
+	return &localImagesGCClient{}
+}
```

Comment:
```suggestion
func defaultImagesGCClient() localImagesGCClient {
	return localImagesGCClient{}
}
```
(we're not holding anything, so we don't need to put this on a pointer)